### ISSUE TITLE
some const values are eagerly optimised away or not reinitialised on scope entry

### ISSUE THREAD

**Issue by catb0t** [trizen/sidef#81]
The nature of this bug is that if you try to demonstrate it in a small example, it no longer appears. 

This bug is the composition of two sane and legitimate optimisations made at the `-O0` 0th level of optimisation:

1.  the value generated for the RHS in `const x = Value()`  is not evaluated until it is eagerly needed directly in the computation of another eager value, or directly needed in the production of a side effect like `say`.

2.  `const` names are elided if the compiler can prove through heuristic that the name is never used, or never directly needed after a certain point.

Combined, these still create a reasonable elision for unused `const` names, whose values are not even computed. 

Currently the compiler incorrectly decides that

1.  a `const` name is eligible for elision if it appears far enough away within the same scope, even as a return value, or 
2.  the name should carry the same value over multiple separate invocations of its scope.

These are so closely related, and worked around in the same way, so they go in the same issue. 

`#1` is demonstrated in https://github.com/catb0t/multifactor/blob/master/guardio.sm#L85-L96
```ruby
module GuardIO {
  # ...
  module Atomic {
    # ...

    # this returns a filehandle to in-memory string
    func aread (File name) -> String {
      const File tmp = Atomic::_temp_r(name)
      const tmp_fh = tmp.open_r
      die "Couldn't open file for reading: #{tmp}: #{$!}" if (! tmp_fh)
      # if declared `var`, it won't be optimised away
      const String contents = tmp_fh.slurp
      die "Read file '#{tmp}' failed: #{$!}" if (! defined(contents))
      contents # won't be eagerly evaluated (will be nil) without this line
      # check error here?
      tmp_fh.close
      # other side-effecting operations...
      contents
    }
    # ... 
  }
  # ...
}
```

Of course, if you use [a simple test file](https://github.com/catb0t/multifactor/blob/master/scratch_scripts/test_guardio.sf) to test this function, it works fine and prints itself, but if it's part of a larger program's longer lifetime and larger generated code, `const contents` is often `nil` by the time it is returned at the end of the scope. 

`#2` is even more complicated to reproduce than that one (sorry!), but at least I can give you an example that works directly. 

https://github.com/catb0t/multifactor/blob/master/ABCs2.sm#L172-L230

```ruby
class GeneratedAccessors {
  method _bound_accessor_generator {
    {
        |field_name, field_rhs|
      func try_index (Num i, Block b = { _ } ) {
        ( TypeUtils::is_a_fqco(field_rhs, Array, must_have_prop: :merge_reduce) )
          ? field_rhs[i]
          : b(field_rhs)
      }
      const typename = try_index(0)
      const default_val = try_index(1, b: { TypeUtils::maybe_call_custom_init( _() ) } )
      # if this is `const`, its value will get mixed up, causing infinite recursion
      var custom_blocks = try_index(2, b: { Hash() } )

      (typename ~~ default_val) \
        || die (
          "#{__FILE__}:#{__LINE__}: #{__METHOD_NAME__}: type mismatch when generating default value for '#{field_name}': ",
          "want '#{typename}'; got '#{default_val.class}' ('#{default_val.dump}')"
      )

      die "class already has property accessor named '#{field_name}' (did you forget to declare a raw property with a leading '_' underscore?)" if ( self.is_property(field_name, default_val) )

      # create the private slot
      say "set _#{field_name} to #{default_val}"
      # if const, custom_blocks should be observed here
      # say "custom_blocks: #{custom_blocks.dump}"
      self{:_ + field_name} = default_val

      const Block setter_prelude = {
        .is_a(typename) \
          || die(
            "#{__FILE__}:#{__LINE__}: #{__METHOD_NAME__}:  ",
            "type mismatch in generated setter '#{field_name}': want '#{typename}'; got '#{_.class}'  ('#{_.dump}')"
        )
      }


      func acc (self) {
        static var custom_getter = custom_blocks{:get}
        say "get field #{field_name} (#{typename}) with #{custom_getter}"
        if (custom_getter) { say "running..."; custom_getter.run(self); say "done" } else { self{:_ + field_name} }
      }

      func acc (self, new_val) {
        static var custom_setter = custom_blocks{:set}
        setter_prelude.run(new_val)
        # say "set field #{field_name} to #{new_val} (#{typename})"
        if (custom_setter) { custom_setter.run(self, new_val) \\ self } else { self{:_ + field_name} = new_val; self }
      }
      __CLASS__.def_method(field_name, acc)
    }
  }

  method init {
    say "#{__CLASS_NAME__}: #{__METHOD_NAME__}"
    self.serial_properties.map_kv( self._bound_accessor_generator )
    self
  }
}
```

`method _bound_accessor_generator` binds `self` in a block used for `map_kv`, and returns that block. 

If you change `var custom_blocks` to `const custom_blocks`, and don't uncomment the `say` where it's used ahead of the `func acc` declarations, [this test script](https://github.com/catb0t/multifactor/blob/master/scratch_scripts/test_abcs2.sf) will recurse forever, despite this code not using any recursion. (And, I can prove this is unrelated to the use of `init`, because changing all constructors to use `.start` does not change the behaviour.) 

From my testing it seems that `const custom_blocks`, once it is initialised to a non-empty hash by [this consuming declaration](https://github.com/catb0t/multifactor/blob/master/builtobj.sm#L28), is not re-initialised the next time the block is called on the next `key, value` pair from `map_kv`. 

This means that sometimes, the explicit accessor `custom_blocks{:get}` for the property `blob_algorithm` is wrongly kept on the next iteration for `blob_sums`'s getter, IFF the `blob_sums` key in `serial_properties` is processed after `blob_algorithm`. If `blob_sums`'s accessor refers to itself, the accessor recurses forever. 

In other words, `blob_sums` has no declared custom getter; its `custom_blocks` should be `Hash()` but sometimes `const custom_blocks`'s value is not re-computed if it's not used early enough.

This also only recurses when `blob_algorithm` is seen first in the Hash iteration, which is not every time, but it is often enough. 

If changed to `var custom_blocks` or `const custom_blocks` but the value is used before the accessor declaration, the code works properly and doesn't recurse.

Hopefully this all makes sense, but I can try to write more code that precisely demonstrates these bugs if necessary.

**Comment by catb0t** [comment]
Wow, that's a simple fix. Thanks!
